perf/bps_tree.cc

```diff
@@ -0,0 +1,464 @@
+generate_benchmark_height(tree_i64, find_last, 4);
+
+template<class tree>
+static void
+test_find_rand(benchmark::State &state, size_t count)
+{
+	test_find<tree>(state, count, RandomKey(count));
+}
+
+generate_benchmark_size(tree_i64, find_rand, 1000000);
+generate_benchmark_size(tree_i64, find_rand, 10000000);
+
+generate_benchmark_height(tree_i64, find_rand, 1);
+generate_benchmark_height(tree_i64, find_rand, 2);
+generate_benchmark_height(tree_i64, find_rand, 3);
+generate_benchmark_height(tree_i64, find_rand, 4);
+
+/*
+ * The following functions test performance of insertion and deletion without
+ * reballancing. This is done by performing the two opposite operations in a
+ * benchmark loop.
+ */
+
+template<class tree, class KeyGen>
+static void
+test_delete_insert(benchmark::State &state, size_t count, KeyGen kg)
+{
+	std::vector<typename tree::elem_t> arr;
+	for (size_t i = 0; i < count; i++)
+		arr.emplace_back(i);
+
+	typename tree::Allocator allocator(count);
+
+	typename tree::tree_t t;
+	tree::create(&t, 0, allocator.alloc, allocator.free,
+		     &allocator, NULL);
+	tree::build(&t, &arr.front(), count);
+	typename tree::elem_t replaced, successor;
+	for (auto _ : state) {
+		typename tree::key_t key(kg());
+		tree::delete_(&t, key);
+		tree::insert(&t, key, &replaced, &successor);
+	}
+	tree::destroy(&t);
+}
+
+template<class tree>
+static void
+test_delete_insert_first(benchmark::State &state, size_t count)
+{
+	test_delete_insert<tree>(state, count, ValueKey(0));
+}
+
+generate_benchmark_size(tree_i64, delete_insert_first, 1000000);
+generate_benchmark_size(tree_i64, delete_insert_first, 10000000);
+
+generate_benchmark_height(tree_i64, delete_insert_first, 1);
+generate_benchmark_height(tree_i64, delete_insert_first, 2);
+generate_benchmark_height(tree_i64, delete_insert_first, 3);
+generate_benchmark_height(tree_i64, delete_insert_first, 4);
+
+template<class tree>
+static void
+test_delete_insert_last(benchmark::State &state, size_t count)
+{
+	test_delete_insert<tree>(state, count, ValueKey(count - 1));
+}
+
+generate_benchmark_size(tree_i64, delete_insert_last, 1000000);
+generate_benchmark_size(tree_i64, delete_insert_last, 10000000);
+
+generate_benchmark_height(tree_i64, delete_insert_last, 1);
+generate_benchmark_height(tree_i64, delete_insert_last, 2);
+generate_benchmark_height(tree_i64, delete_insert_last, 3);
+generate_benchmark_height(tree_i64, delete_insert_last, 4);
+
+template<class tree>
+static void
+test_delete_insert_rand(benchmark::State &state, size_t count)
+{
+	test_delete_insert<tree>(state, count, RandomKey(count));
+}
+
+generate_benchmark_size(tree_i64, delete_insert_rand, 1000000);
+generate_benchmark_size(tree_i64, delete_insert_rand, 10000000);
+
+generate_benchmark_height(tree_i64, delete_insert_rand, 1);
+generate_benchmark_height(tree_i64, delete_insert_rand, 2);
+generate_benchmark_height(tree_i64, delete_insert_rand, 3);
+generate_benchmark_height(tree_i64, delete_insert_rand, 4);
+
```

Comment:
What about insertion/deletion of successive keys for cache coherence?

Reply:
Seems useful since we have different lookup speed for first and last element in a block (for some reason). The new benchmarks may show a smoothed performance over all the tree.

Added two more `delete_insert` and `find` benchmarks.